### PR TITLE
PXC-3743: Make wsrep session variables updatable through SET_VAR() hint

### PR DESCRIPTION
**Pull request by venkatesh-prasad-v**
https://jira.percona.com/browse/PXC-3743

Allow users to utilize SET_VAR optimizer hint to temporarily set the wsrep
session variables.

This patch allows the following session variables to be updatable through
SET_VAR() extension.

1. wsrep_sync_wait
2. wsrep_dirty_reads
3. wsrep_trx_fragment_size
4. wsrep_trx_fragment_unit
```
Example:

mysql> SELECT @@wsrep_sync_wait; SELECT /*+ SET_VAR(wsrep_sync_wait=4 )*/ @@wsrep_sync_wait; SELECT @@wsrep_sync_wait;
+-------------------+
| @@wsrep_sync_wait |
+-------------------+
|                15 |
+-------------------+
1 row in set (0.00 sec)
+-------------------+
| @@wsrep_sync_wait |
+-------------------+
|                 4 |
+-------------------+
1 row in set (0.00 sec)
+-------------------+
| @@wsrep_sync_wait |
+-------------------+
|                15 |
+-------------------+
1 row in set (0.00 sec)
```
Note: The following session variables cannot be updatable through
SET_VAR

1. wsrep_on: Because it may cause inconsistent read/writes.
2. wsrep_causal_reads: Deprecated.

In addition to the above change, the `wsrep` suite has been added to the
default MTR suite as well.

Testing Done
---
Galera suite: https://pxc.cd.percona.com/view/PXC%208.0/job/pxc-8.0-param/364/testReport/
Failing tests: 

1. galera.pxc3444 - Fixed in [8.0.26 merge branch commit](https://github.com/percona/percona-xtradb-cluster/commit/70e244526e6f5ad170cbccb91a5e16488bd3d531#diff-a7a6d5fdbce4f8d0148784d04113fc252647784edbda2485d4e63323647f2191)
2. galera.pxc3733 - Fixed in [8.0.26 merge branch commit](https://github.com/percona/percona-xtradb-cluster/commit/70e244526e6f5ad170cbccb91a5e16488bd3d531#diff-547dd82845ced28a26bae75ec044c217ab5e64ccb4dcf2873f4de51af18bfd9d)
3. galera.pxc_strict_mode - Fixed in [8.0.26 merge branch commit](https://github.com/percona/percona-xtradb-cluster/commit/70e244526e6f5ad170cbccb91a5e16488bd3d531#diff-2e34c8537043419fde5b741f5a3dbd1aff91bce0b0d24fa28f67872df4b818c5)
4. galera.GCF-360 - Fixed in [8.0.26 merge branch commit](https://github.com/percona/percona-xtradb-cluster/commit/70e244526e6f5ad170cbccb91a5e16488bd3d531#diff-89ec7cddddec219c590d0251ea6dc567ad4156cc679554511e47ec0346084322)

### CODE REVIEW

File: mysql-test/suite/wsrep/my.cnf [hunk 1]
```diff
@@ -2,10 +2,17 @@
...
+innodb_flush_log_at_trx_commit=2
+
+# wsrep variables
+wsrep_causal_reads=ON
 wsrep_cluster_address='gcomm://'
 wsrep_node_address=127.0.0.1
-binlog-format=row
+wsrep_node_name=node1
+wsrep_provider=@ENV.WSREP_PROVIDER
+wsrep_sync_wait = 15
```
Comment (kamil-holubicki) [inline mysql-test/suite/wsrep/my.cnf]: Inconsistent formating

File: mysql-test/suite/wsrep/t/wsrep_set_var_extension.test [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+
+DELIMITER ;\\
+
+--echo #########################################
+--echo # Test SET_VAR hint for wsrep variables #
+--echo #########################################
+
+--echo
+--echo # Test 1: wsrep_sync_wait
+CALL test_hint("SET_VAR(wsrep_sync_wait=4)", "wsrep_sync_wait");
```
Comment (kamil-holubicki) [inline mysql-test/suite/wsrep/t/wsrep_set_var_extension.test:55]: What is the purpose of this call?

File: mysql-test/suite/wsrep/t/wsrep_set_var_extension.test [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+
+SET wsrep_sync_wait=2;
+CALL test_hint("SET_VAR(wsrep_sync_wait=4)", "wsrep_sync_wait");
+
+--assert (`SELECT @before_val != @hint_val`)
+--assert (`SELECT @hint_val != @after_val`)
+--assert (`SELECT @before_val = @after_val`)
+
+--assert (`SELECT /*+ SET_VAR(wsrep_sync_wait=10)*/ @@wsrep_sync_wait = 10`)
```
Comment (kamil-holubicki) [inline mysql-test/suite/wsrep/t/wsrep_set_var_extension.test]: Wouldn't it be more clear to add 1 more assert to 3 above asserts like
`--assert (SELECT @hint_val = 4)`
?

File: mysql-test/suite/wsrep/t/wsrep_set_var_extension.test [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+SET wsrep_sync_wait=default;
+
+--echo
+--echo # Test 2: wsrep_dirty_reads
+CALL test_hint("SET_VAR(wsrep_dirty_reads=1)", "wsrep_dirty_reads");
```
Comment (kamil-holubicki) [inline mysql-test/suite/wsrep/t/wsrep_set_var_extension.test]: As above, what is the purpose of this first CALL?

File: mysql-test/suite/wsrep/t/wsrep_set_var_extension.test [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+--assert (`SELECT @before_val != @hint_val`)
+--assert (`SELECT @hint_val != @after_val`)
+--assert (`SELECT @before_val = @after_val`)
+
+--assert (`SELECT /*+ SET_VAR(wsrep_dirty_reads=1)*/ @@wsrep_dirty_reads = 1`)
+SET wsrep_dirty_reads=default;
+
+
+--echo
+--echo # Test 2: wsrep_dirty_reads
```
Comment (kamil-holubicki) [inline mysql-test/suite/wsrep/t/wsrep_set_var_extension.test]: Does this check/test differ from the one above?

**Review by kamil-holubicki — COMMENTED** [review]
The change looks fine.
I left a couple of questions, suggestions.

File: mysql-test/suite/wsrep/t/wsrep_set_var_extension.test (hunk 2 above)
Comment (venkatesh-prasad-v) [reply mysql-test/suite/wsrep/t/wsrep_set_var_extension.test:55]: The intention here is to first test with the MTR's default value for the variables, then changing the value and testing again.

File: mysql-test/suite/wsrep/my.cnf (hunk 1 above)
Comment (venkatesh-prasad-v) [reply mysql-test/suite/wsrep/my.cnf]: Fixed.

File: mysql-test/suite/wsrep/t/wsrep_set_var_extension.test (hunk 5 above)
Comment (venkatesh-prasad-v) [reply mysql-test/suite/wsrep/t/wsrep_set_var_extension.test]: Thanks for catching this. This is a duplicate. Removed now.

File: mysql-test/suite/wsrep/t/wsrep_set_var_extension.test (hunk 3 above)
Comment (venkatesh-prasad-v) [reply mysql-test/suite/wsrep/t/wsrep_set_var_extension.test]: Added.

File: mysql-test/suite/wsrep/t/wsrep_set_var_extension.test (hunk 2 above)
Comment (kamil-holubicki) [reply mysql-test/suite/wsrep/t/wsrep_set_var_extension.test:55]: OK, but shouldn't we do asserts as below? The stored procedure does not do any validation.

File: mysql-test/suite/wsrep/t/wsrep_set_var_extension.test (hunk 3 above)
Comment (kamil-holubicki) [reply mysql-test/suite/wsrep/t/wsrep_set_var_extension.test]: My intention was to add this assert and remove this:
`--assert (SELECT /*+ SET_VAR(wsrep_sync_wait=10)*/ @@wsrep_sync_wait = 10)`
Makes sense?

File: mysql-test/suite/wsrep/t/wsrep_set_var_extension.test (hunk 2 above)
Comment (venkatesh-prasad-v) [reply mysql-test/suite/wsrep/t/wsrep_set_var_extension.test:55]: We want to stop executing the test immediately our expectations are not met and this is not possible with the stored routines. So, the stored procedure only sets the uservars and they will be later used for the validation.

File: mysql-test/suite/wsrep/t/wsrep_set_var_extension.test (hunk 3 above)
Comment (venkatesh-prasad-v) [reply mysql-test/suite/wsrep/t/wsrep_set_var_extension.test]: Done. Removed.

File: mysql-test/suite/wsrep/t/wsrep_set_var_extension.test (hunk 2 above)
Comment (venkatesh-prasad-v) [reply mysql-test/suite/wsrep/t/wsrep_set_var_extension.test:55]: I think I got your point now. You meant to add asserts even after the first call right?

File: mysql-test/suite/wsrep/my.cnf [hunk 6]
```diff
@@ -2,10 +2,17 @@
...
+innodb_flush_log_at_trx_commit=2
+
+# wsrep variables
+wsrep_causal_reads=ON
 wsrep_cluster_address='gcomm://'
 wsrep_node_address=127.0.0.1
-binlog-format=row
+wsrep_node_name=node1
+wsrep_provider=@ENV.WSREP_PROVIDER
+wsrep_sync_wait=15
```
Comment (satya-bodapati) [inline mysql-test/suite/wsrep/my.cnf]: why did you add the wsrep_sync_wait change to suite cnf? wouldn't it affect entire suite?

Comment (venkatesh-prasad-v) [reply mysql-test/suite/wsrep/my.cnf]: This is the first test to go into the `wsrep` suite in 8.0. So, we are safe :)

Comment (venkatesh-prasad-v) [reply mysql-test/suite/wsrep/my.cnf]: The reason for adding it to the cnf is to get consistent reads in the test cases so that it becomes easier to write tests  without thinking about synchronization of previous writes.
```
--connection node_1
INSERT INTO t1 VALUES(1);

--connection node_2
SELECT * FROM t1;
```
With wsrep_sync_wait set to 15, we can always ensure that `SELECT` on `node_2` is executed only after `INSERT` is replicated and applied on `node_2`

Comment (satya-bodapati) [reply mysql-test/suite/wsrep/my.cnf]: ok, 15 comes from this. So we do cluster-sync check on these SQLs.
```
enum enum_wsrep_sync_wait {
    WSREP_SYNC_WAIT_NONE = 0x0,
    // select, begin
    WSREP_SYNC_WAIT_BEFORE_READ = 0x1,
    WSREP_SYNC_WAIT_BEFORE_UPDATE_DELETE = 0x2,
    WSREP_SYNC_WAIT_BEFORE_INSERT_REPLACE = 0x4,
    WSREP_SYNC_WAIT_BEFORE_SHOW = 0x8,
    WSREP_SYNC_WAIT_MAX = 0xF
  };
```

Comment (satya-bodapati) [reply mysql-test/suite/wsrep/my.cnf]: @venkatesh-prasad-v if this is a new suite, I would rename it to wsrep_one_node or wsrep_1node

Comment (satya-bodapati) [reply mysql-test/suite/wsrep/my.cnf]: @venkatesh-prasad-v do you know why we are different from upstream for this variable? upstream supports an additional value of 3.
https://galeracluster.com/library/documentation/mysql-wsrep-options.html#wsrep-sync-wait

File: mysql-test/suite/wsrep/t/wsrep_set_var_extension.test (hunk 2 above)
Comment (kamil-holubicki) [reply mysql-test/suite/wsrep/t/wsrep_set_var_extension.test:55]: Exactly

File: mysql-test/suite/wsrep/my.cnf (hunk 6 above)
Comment (venkatesh-prasad-v) [reply mysql-test/suite/wsrep/my.cnf]: @satya-bodapati 
> if this is a new suite, I would rename it to wsrep_one_node or wsrep_1node

wsrep suite is actually meant for one node only. So it is implicit and there is no need to add any extension.


>do you know why we are different from upstream for this variable? upstream supports an additional value of 3.

We also support value 3 (since it is bitmask internally) . It has also been documented well in https://www.percona.com/doc/percona-xtradb-cluster/8.0/wsrep-system-index.html#wsrep_sync_wait

```
mysql> set wsrep_sync_wait=3;
Query OK, 0 rows affected (0.00 sec)

mysql> select @@wsrep_sync_wait;
+-------------------+
| @@wsrep_sync_wait |
+-------------------+
|                 3 |
+-------------------+
1 row in set (0.00 sec)
```

File: mysql-test/suite/wsrep/t/wsrep_set_var_extension.test (hunk 2 above)
Comment (venkatesh-prasad-v) [reply mysql-test/suite/wsrep/t/wsrep_set_var_extension.test:55]: Fixed now.

File: mysql-test/suite/wsrep/my.cnf (hunk 6 above)
Comment (kamil-holubicki) [reply mysql-test/suite/wsrep/my.cnf]: >With wsrep_sync_wait set to 15, we can always ensure that SELECT on node_2 is executed only after INSERT is replicated and applied on node_2

Wasn't the intention of this suite to use only 1 node?

>if this is a new suite, I would rename it to wsrep_one_node or wsrep_1node

or galera_1node :)

**Review by satya-bodapati — APPROVED** [review]
LGTM!
